### PR TITLE
Update `createContext` to memoize props internally

### PR DESCRIPTION
**Pull request by jjenzz**
You can test how this behaves here https://codesandbox.io/s/aged-darkness-wlkxr?file=/src/App.tsx

### CODE REVIEW

File: packages/react/utils/src/createContext.tsx [hunk 1]
```diff
@@ -18,13 +18,13 @@ export function createContextObj<ContextValueType extends object>(rootComponentN
...
-      () => providerProps,
-      // eslint-disable-next-line react-hooks/exhaustive-deps
-      Object.values(providerProps)
-    ) as ContextValueType;
-    return <Context.Provider value={value}>{children}</Context.Provider>;
+    const providerPropsString = JSON.stringify(providerProps);
+    // Only re-memoize when stringified props change
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const context = React.useMemo(() => ({} as ContextValueType), [providerPropsString]);
+    // mutate properties to ensure non primitive values are up to date
```
Comment (jjenzz) [inline packages/react/utils/src/createContext.tsx:25]:
```suggestion
    // mutate properties to ensure closures are up to date
```

File: packages/react/utils/src/createContext.tsx [hunk 2]
```diff
@@ -18,13 +18,13 @@ export function createContextObj<ContextValueType extends object>(rootComponentN
...
-      // eslint-disable-next-line react-hooks/exhaustive-deps
-      Object.values(providerProps)
-    ) as ContextValueType;
-    return <Context.Provider value={value}>{children}</Context.Provider>;
+    const providerPropsString = JSON.stringify(providerProps);
+    // Only re-memoize when stringified props change
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const context = React.useMemo(() => ({} as ContextValueType), [providerPropsString]);
+    // mutate properties to ensure closures are up to date
+    Object.entries(providerProps).forEach(([prop, value]) => ((context as any)[prop] = value));
```
Comment (benoitgrelard) [inline packages/react/utils/src/createContext.tsx:26]: I think I'm gonna need some help understanding how this actually works.

**Review by benoitgrelard — COMMENTED** [review]
We have some build failures because some things passed in are sometimes react elements, so don't stringify.

I'm also keen to get on a call to go through this as I'm not sure I fully understand how this works.

File: packages/react/utils/src/createContext.tsx (hunk 2 above)
Comment (jjenzz) [reply packages/react/utils/src/createContext.tsx:26]: Ah I think I need to do that `JSON.stringify` filter thing we did with Jonathan the other day. Perhaps there is another way to achieve this without `JSON.stringify` though. Let's have a call when you're here so I can explain what's needed here

**Comment by jjenzz** [comment]
We've both agreed this is getting a bit out of hand, so we're gonna keep with the inline `React.useCallback` approach.

**Pull request closed without merging**